### PR TITLE
fix Animation reverse wrong effect, loop animation memory leak

### PR DESCRIPTION
**Pull request by drelaptop**
see the WrapMode defined in creator:
```javascript
var WrapModeMask = {
    Loop: 1 << 1,
    ShouldWrap: 1 << 2,
    // Reserved: 1 << 3,
    PingPong: 1 << 4 | 1 << 1 | 1 << 2,  // Loop, ShouldWrap
    Reverse: 1 << 5 | 1 << 2,      // ShouldWrap
};
```
refer to the source code: https://github.com/cocos-creator/engine/blob/master/cocos2d/animation/types.js#L28

modified the enum values, keep same as the creator defined, issue #151 solved

### CODE REVIEW

**Comment by drelaptop** [comment]
when I test the wrap mode one by one, the `LoopReverse ` hold the same behavior with `PingPongReverse `, I am trying to solve this, other WrapMode works well.

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp [hunk 1]
```diff
@@ -205,19 +205,20 @@ bool AnimateClip::initWithAnimationClip(cocos2d::Node* rootTarget, AnimationClip
...
-        
+
         auto wrapMode = clip->getWrapMode();
         if (wrapMode == AnimationClip::WrapMode::Loop || wrapMode == AnimationClip::WrapMode::LoopReverse)
             _needStop = false;
-        
+
         _durationToStop = _clip->getDuration();
         if (wrapMode == AnimationClip::WrapMode::PingPong || wrapMode == AnimationClip::WrapMode::PingPongReverse)
+            _needStop = false;
```
Comment (drelaptop) [inline creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp]: `PingPong` and `PingPongReverse` always keep Loop, refer to the [creator docs](http://docs.cocos.com/creator/api/zh/enums/WrapMode.html)

> PingPong
Even iterations are played as specified, odd iterations are played in the reverse direction from the way they are specified.
从第一帧播放到最后一帧，然后反向播放回第一帧，到第一帧后再正向播放，如此循环

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp [hunk 2]
```diff
@@ -319,19 +320,19 @@ float AnimateClip::computeElapse() const
 {
     auto elapsed = _elapsed;
     auto duration = _clip->getDuration();
-    
-    // if wrap mode is pingpong or pingpongreverse, then _elapsed may be bigger than duration
-    while (elapsed >= duration)
-        elapsed = elapsed - duration;
```
Comment (drelaptop) [inline creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp]: while loop will be executed many times as the time goes

**Comment by drelaptop** [comment]
@minggo please review and merge

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp [hunk 3]
```diff
@@ -319,19 +320,19 @@ float AnimateClip::computeElapse() const
...
-    
+
+    // as the time goes, _elapsed will be bigger than duration when _needStop = false
+    elapsed  = elapsed - static_cast<int>(elapsed / duration) * duration;
```
Comment (minggo) [inline creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp]: Yep, the code is better, but how about
```c++
elapsed  %= duration;
```
?

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp [hunk 4]
```diff
@@ -319,19 +320,19 @@ float AnimateClip::computeElapse() const
...
+
     const auto wrapMode = _clip->getWrapMode();
     bool oddRound = (static_cast<int>(_elapsed / duration) % 2) == 0;
     if (wrapMode == AnimationClip::WrapMode::Reverse  // reverse mode
         || (wrapMode == AnimationClip::WrapMode::PingPong && !oddRound) // pingpong mode and it is the second round
         || (wrapMode == AnimationClip::WrapMode::PingPongReverse && oddRound) // pingpongreverse mode and it is the first round
-        || (wrapMode == AnimationClip::WrapMode::LoopReverse && oddRound)
+        || (wrapMode == AnimationClip::WrapMode::LoopReverse)  // loop reverse mode, reverse again and again
```
Comment (minggo) [inline creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp]: Why remove `oddRound`?

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp (hunk 3 above)
Comment (drelaptop) [reply creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp]: Both are float number, I can use the `fmodf` function

**Comment by drelaptop** [comment]
please merge @minggo

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp [hunk 5]
```diff
@@ -227,16 +234,14 @@ bool AnimateClip::initWithAnimationClip(cocos2d::Node* rootTarget, AnimationClip
 
 void AnimateClip::update(float dt) {
     _elapsed += dt;
-    
+
     if (_needStop && _elapsed >= _durationToStop)
     {
         stopAnimate();
-        if (_endCallback)
-            _endCallback();
```
[the remark this answers is not in the document]
Comment (drelaptop) [reply creator_project/packages/creator-luacpp-support/reader/animation/AnimateClip.cpp]: Remove it here, and add it into `stopAnimate()`
```C++
void AnimateClip::stopAnimate()
{
    if (_endCallback)
        _endCallback();

    unscheduleUpdate();
    // release self
    _running = false;
}
```
for the meaning `_endCallback` is a stop action, so I can move it into `stopAnimate()`, another reason is to reduce duplicate, I used `stopAnimate` twice, another is:
```
AnimateClip::~AnimateClip()
{
    // a loop animate might keep running until destruction, memory will leak if not stop it
    if(_running)
        stopAnimate();

    CC_SAFE_RELEASE(_clip);
    CC_SAFE_RELEASE(_rootTarget);
}```

**Comment by nunu-e64** [comment]
I read all commits in this PR.
LGTM.

File: creator_project/packages/creator-luacpp-support/reader/animation/AnimationManager.cpp [hunk 6]
```diff
@@ -27,7 +27,7 @@ void AnimationManager::playOnLoad()
     }
 }
 
-void AnimationManager::stopPlayOnLoad()
+void AnimationManager::stopAnimationClipRunByPlayOnLoad()
```
Comment (minggo) [inline creator_project/packages/creator-luacpp-support/reader/animation/AnimationManager.cpp]: stopAnimationClipRunByPlayOnLoad -> stopAnimationClipsRunByPlayOnLoad

Comment (drelaptop) [reply creator_project/packages/creator-luacpp-support/reader/animation/AnimationManager.cpp]: thanks, done

**Pull request merged**